File: app/crypto.py
"""
DFARS Desktop - encryption key management for secrets at rest.

Used by auth.py to encrypt the user's TOTP secret before writing it to
auth.db. Without encryption, anyone who copies auth.db gets the MFA
seed; with it, they also need the Fernet key.

Key storage (in order of preference):

1. Keyring (Windows Credential Manager via DPAPI, macOS Keychain, or
   Secret Service on Linux). This ties the key to the current OS user
   account — moving the data to another user or machine breaks decryption,
   which is exactly what we want for a forensic tool.

2. File fallback at %APPDATA%\\DFARS\\.keyfile if keyring is unavailable.
   This is strictly worse (the file sits next to auth.db) but ensures
   the app still works in containers, WSL, minimal environments, etc.
   The file is created with 0o600 permissions on POSIX; on Windows the
   ACLs come from inheritance and there's no portable way to restrict
   further from Python.

A lost key means encrypted TOTP secrets become unrecoverable, but the
user's **recovery codes** (stored as separate hashes in auth.db) still
work, so they can always log in and re-enroll MFA.
"""

from __future__ import annotations

import logging
import os
from pathlib import Path
from typing import Optional

from cryptography.fernet import Fernet, InvalidToken

from .paths import data_dir, ensure_data_tree

log = logging.getLogger(__name__)

_KEYRING_SERVICE = "DFARS Desktop"
_KEYRING_USERNAME = "totp_encryption_key"
_KEYFILE_NAME = ".keyfile"


# ─── Key acquisition ───────────────────────────────────────


def _keyfile_path() -> Path:
    return data_dir() / _KEYFILE_NAME


def _try_keyring_get() -> Optional[str]:
    """Read the Fernet key from the OS keyring. Returns None on any failure."""
    try:
        import keyring
        value = keyring.get_password(_KEYRING_SERVICE, _KEYRING_USERNAME)
        return value
    except Exception as e:
        log.debug("Keyring get failed: %s", e)
        return None


def _try_keyring_set(key: str) -> bool:
    """Write the Fernet key to the OS keyring. Returns True on success."""
    try:
        import keyring
        keyring.set_password(_KEYRING_SERVICE, _KEYRING_USERNAME, key)
        return True
    except Exception as e:
        log.debug("Keyring set failed: %s", e)
        return False


def _read_keyfile() -> Optional[str]:
    path = _keyfile_path()
    if not path.exists():
        return None
    try:
        return path.read_text(encoding="ascii").strip()
    except OSError as e:
        log.warning("Failed to read keyfile: %s", e)
        return None


def _write_keyfile(key: str) -> None:
    ensure_data_tree()
    path = _keyfile_path()
    path.write_text(key, encoding="ascii")
    try:
        os.chmod(path, 0o600)
    except (OSError, NotImplementedError):
        # Windows doesn't honour chmod portably; ACLs come from parent dir.
        pass


def _get_or_create_key() -> bytes:
    """
    Return the Fernet key as bytes, creating one on first use.
    Tries keyring first, then falls back to a keyfile in AppData.
    """
    # 1. Try keyring
    key = _try_keyring_get()
    if key:
        return key.encode("ascii")

    # 2. Try keyfile fallback
    key = _read_keyfile()
    if key:
        log.info("Loaded crypto key from keyfile fallback")
        return key.encode("ascii")

    # 3. Generate a new key, prefer keyring for storage, fall back to file
    new_key = Fernet.generate_key().decode("ascii")
    if _try_keyring_set(new_key):
        log.info("Generated new crypto key, stored in OS keyring")
    else:
        _write_keyfile(new_key)
        log.warning(
            "Keyring unavailable; stored crypto key in %s. "
            "This is less secure than keyring — see docs.",
            _keyfile_path(),
        )
    return new_key.encode("ascii")


# ─── Public API ────────────────────────────────────────────


def encrypt(plaintext: bytes | str) -> str:
    """Encrypt and return a URL-safe base64 ciphertext string."""
    if isinstance(plaintext, str):
        plaintext = plaintext.encode("utf-8")
    f = Fernet(_get_or_create_key())
    return f.encrypt(plaintext).decode("ascii")


def decrypt(ciphertext: str) -> bytes:
    """
    Decrypt a Fernet ciphertext string. Raises InvalidToken if the key
    has been rotated/lost or the ciphertext was tampered with.
    """
    f = Fernet(_get_or_create_key())
    return f.decrypt(ciphertext.encode("ascii"))


def try_decrypt(ciphertext: str) -> Optional[bytes]:
    """
    Non-raising decrypt: returns None on InvalidToken. Used by MFA flows
    that should degrade to 'recovery codes only' if the key is missing.
    """
    try:
        return decrypt(ciphertext)
    except InvalidToken:
        log.warning(
            "Failed to decrypt stored ciphertext — key may have been "
            "rotated or lost. User must re-enroll MFA."
        )
        return None

File: v2/src-tauri/src/crypto.rs
/// Fernet symmetric encryption wrapper.
///
/// Manages one Fernet key that protects all secrets-at-rest:
///   - TOTP secrets (`users.totp_secret`)
///   - Agent Zero API key (`config.json → agent_zero_api_key_encrypted`)
///   - SMTP password  (`config.json → smtp_password_encrypted`)
///
/// Key acquisition order (matching v1 `app/crypto.py`):
///   1. Windows Credential Manager via `keyring`, service = "DFARS Desktop",
///      account = "totp_encryption_key"  — exact strings, case-sensitive.
///   2. File fallback: `%APPDATA%\DFARS\.keyfile`
///   3. Generate a new key, write to keyring first; fall back to keyfile if
///      keyring is unavailable.
///
/// MUST-DO 1 (SEC-1): These service/account names are non-negotiable.
/// Using any other name silently orphans all v1 encrypted data.
use std::path::PathBuf;

use tracing::{info, warn};
use zeroize::Zeroizing;

use crate::error::AppError;

// ─── Keyring names — DO NOT CHANGE ────────────────────────────────────────────
// These are the exact strings v1's crypto.py uses in Windows Credential Manager.
// Any deviation silently creates a new entry and renders all v1 encrypted data
// (TOTP secrets, Agent Zero API key, SMTP password) permanently unreadable.
const KEYRING_SERVICE: &str = "DFARS Desktop";
const KEYRING_ACCOUNT: &str = "totp_encryption_key";

// ─── State carried for the lifetime of the process ─────────────────────────

/// Identifies where the Fernet key came from. Returned by
/// `settings_get_security_posture()` so the UI can warn the user if they're
/// running with the less-secure file fallback (SEC-1 SHOULD-DO 6).
#[derive(Debug, Clone, serde::Serialize)]
#[serde(rename_all = "snake_case")]
pub enum KeySource {
    Keyring,
    Keyfile,
    New,
}

pub struct CryptoState {
    /// The Fernet instance. Held in a `Zeroizing` wrapper so the key bytes
    /// are wiped from memory on drop.
    fernet: fernet::Fernet,
    /// Where the key came from — exposed to the security posture command.
    pub key_source: KeySource,
}

// ─── Key acquisition ────────────────────────────────────────────────────────

fn keyfile_path() -> Result<PathBuf, AppError> {
    // Use the `directories` crate to resolve %APPDATA% portably.
    let dirs = directories::BaseDirs::new()
        .ok_or_else(|| AppError::Io("could not resolve base directories".into()))?;
    // %APPDATA%\DFARS\.keyfile — matches v1's `data_dir() / ".keyfile"` exactly.
    Ok(dirs.data_dir().join("DFARS").join(".keyfile"))
}

fn try_keyring_get() -> Option<Zeroizing<String>> {
    let entry = keyring::Entry::new(KEYRING_SERVICE, KEYRING_ACCOUNT).ok()?;
    match entry.get_password() {
        Ok(pw) => Some(Zeroizing::new(pw)),
        Err(e) => {
            tracing::debug!("keyring get failed: {e}");
            None
        }
    }
}

fn try_keyring_set(key: &str) -> bool {
    match keyring::Entry::new(KEYRING_SERVICE, KEYRING_ACCOUNT) {
        Ok(entry) => match entry.set_password(key) {
            Ok(()) => true,
            Err(e) => {
                tracing::debug!("keyring set failed: {e}");
                false
            }
        },
        Err(e) => {
            tracing::debug!("keyring entry creation failed: {e}");
            false
        }
    }
}

fn read_keyfile() -> Option<Zeroizing<String>> {
    let path = keyfile_path().ok()?;
    if !path.exists() {
        return None;
    }
    match std::fs::read_to_string(&path) {
        Ok(s) => Some(Zeroizing::new(s.trim().to_owned())),
        Err(e) => {
            warn!("failed to read keyfile: {e}");
            None
        }
    }
}

fn write_keyfile(key: &str) -> Result<(), AppError> {
    let path = keyfile_path()?;
    if let Some(parent) = path.parent() {
        std::fs::create_dir_all(parent)?;
    }
    std::fs::write(&path, key)?;
    Ok(())
}

fn build_fernet(key_str: &str) -> Result<fernet::Fernet, AppError> {
    fernet::Fernet::new(key_str)
        .ok_or_else(|| AppError::Crypto("invalid Fernet key format".into()))
}

// ─── Public init ─────────────────────────────────────────────────────────────

/// Initialise the crypto layer.  Called once at startup inside the Tauri
/// `.setup()` closure.  Returns a `CryptoState` to be placed into `AppState`.
///
/// Logs the key source at INFO level — NEVER logs the key value.
pub fn init() -> Result<CryptoState, AppError> {
    // 1. Try Windows Credential Manager.
    if let Some(key) = try_keyring_get() {
        info!(key_source = "keyring", "Fernet key loaded");
        let fernet = build_fernet(&key)?;
        return Ok(CryptoState {
            fernet,
            key_source: KeySource::Keyring,
        });
    }

    // 2. Try keyfile fallback (%APPDATA%\DFARS\.keyfile).
    if let Some(key) = read_keyfile() {
        info!(key_source = "keyfile", "Fernet key loaded from file fallback");
        let fernet = build_fernet(&key)?;
        return Ok(CryptoState {
            fernet,
            key_source: KeySource::Keyfile,
        });
    }

    // 3. Generate a new key and persist it.
    let new_key = fernet::Fernet::generate_key();
    // Prefer keyring; fall back to file.
    if try_keyring_set(&new_key) {
        info!(key_source = "new", "Generated new Fernet key, stored in keyring");
        // Also write keyfile as a backup.
        if let Err(e) = write_keyfile(&new_key) {
            warn!("could not write keyfile backup: {e}");
        }
    } else {
        warn!(
            key_source = "new",
            "Keyring unavailable; storing Fernet key in keyfile. \
             This is less secure — consider enabling Windows Credential Manager."
        );
        write_keyfile(&new_key)?;
    }

    let fernet = build_fernet(&new_key)?;
    Ok(CryptoState {
        fernet,
        key_source: KeySource::New,
    })
}

// ─── Public encrypt / decrypt ─────────────────────────────────────────────

impl CryptoState {
    /// Encrypt arbitrary bytes and return a URL-safe Fernet token string.
    pub fn encrypt(&self, plaintext: &[u8]) -> String {
        self.fernet.encrypt(plaintext)
    }

    /// Decrypt a Fernet token string back to bytes.
    pub fn decrypt(&self, ciphertext: &str) -> Result<Vec<u8>, AppError> {
        self.fernet
            .decrypt(ciphertext)
            .map_err(|_| AppError::Crypto("decryption failed — key may have been rotated or lost".into()))
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    /// Round-trip: bytes encrypted by this module must decrypt to the original.
    #[test]
    fn encrypt_decrypt_roundtrip() {
        // Use a known valid Fernet key (32 bytes URL-safe-base64-encoded).
        let key = fernet::Fernet::generate_key();
        let fernet_inst = fernet::Fernet::new(&key).unwrap();
        let state = CryptoState {
            fernet: fernet_inst,
            key_source: KeySource::New,
        };
        let plaintext = b"dfars-test-secret-totp-seed";
        let ct = state.encrypt(plaintext);
        let recovered = state.decrypt(&ct).unwrap();
        assert_eq!(recovered, plaintext);
    }

    /// Wrong key must return an error, not panic.
    #[test]
    fn decrypt_wrong_key_returns_error() {
        let key1 = fernet::Fernet::generate_key();
        let key2 = fernet::Fernet::generate_key();
        let state1 = CryptoState {
            fernet: fernet::Fernet::new(&key1).unwrap(),
            key_source: KeySource::New,
        };
        let state2 = CryptoState {
            fernet: fernet::Fernet::new(&key2).unwrap(),
            key_source: KeySource::New,
        };
        let ct = state1.encrypt(b"secret");
        assert!(state2.decrypt(&ct).is_err());
    }
}
